Mathlib/RingTheory/Spectrum/Prime/Homeomorph.lean

```diff
@@ -0,0 +1,107 @@
+import Mathlib.RingTheory.Spectrum.Prime.Topology
+
+/-!
+# Purely inseparable extensions are universal homeomorphisms
+
+If `K` is a purely inseparable extension of `k`, the induced map `Spec K ⟶ Spec k` is a universal
+homeomorphism, i.e. it stays a homeomorphism after arbitrary base change.
+
+## Main results
+
+- `PrimeSpectrum.isHomeomorph_comap`: if `f : R →+* S` is a ring map with locally nilpotent kernel
+  such that for every `x : S`, there exists `n > 0` such that `x ^ n` is in the image of `f`,
+  `Spec f` is a homeomorphism.
+- `PrimeSpectrum.isHomeomorph_comap_of_isPurelyInseparable`: `Spec K ⟶ Spec k` is a universal
+  homeomorphism for a purely inseparable field extension `K` over `k`.
+-/
+
+open TensorProduct
+
+variable (k K R S : Type*) [Field k] [Field K] [Algebra k K] [CommRing R] [Algebra k R] [CommRing S]
+
+variable {R S} in
+@[stacks 0BR8 "Homeomorphism part"]
+lemma PrimeSpectrum.isHomeomorph_comap (f : R →+* S) (H : ∀ (x : S), ∃ n > 0, x ^ n ∈ f.range)
+    (hker : RingHom.ker f ≤ nilradical R) : IsHomeomorph (PrimeSpectrum.comap f) := by
+  have h1 : Function.Injective (PrimeSpectrum.comap f) := by
+    intro q q' hqq'
+    ext x
+    by_contra! h
+    wlog ha : x ∈ q.asIdeal ∧ x ∉ q'.asIdeal generalizing q q'
+    · exact this hqq'.symm (Or.inl <| by tauto) (by tauto)
+    obtain ⟨hxq, hxq'⟩ := ha
+    obtain ⟨n, hn, y, hy⟩ := H x
+    simp only [PrimeSpectrum.ext_iff, SetLike.ext'_iff, PrimeSpectrum.comap_asIdeal,
+      Ideal.coe_comap] at hqq'
+    have : x ^ n ∈ q'.asIdeal := by
+      rw [← hy, ← SetLike.mem_coe, ← Set.mem_preimage, ← hqq', Set.mem_preimage, hy]
+      exact Ideal.pow_mem_of_mem q.asIdeal hxq n hn
+    exact hxq' (q'.2.mem_of_pow_mem n this)
+  have hint : f.kerLift.IsIntegral := by
+    intro x
+    obtain ⟨n, hn, y, hy⟩ := H x
+    use Polynomial.X ^ n - Polynomial.C (Ideal.Quotient.mk _ y)
+    simp only [Polynomial.eval₂_sub, Polynomial.eval₂_X_pow, ← hy, Polynomial.eval₂_C,
+      RingHom.kerLift_mk, sub_self, and_true]
+    apply Polynomial.monic_X_pow_add
+    simpa using lt_of_le_of_lt Polynomial.degree_C_le (by simpa using hn)
+  have hbij : Function.Bijective (PrimeSpectrum.comap f) :=
+    ⟨h1, (PrimeSpectrum.comap_quotientMk_bijective_of_le_nilradical _ hker).2.comp <|
+      hint.specComap_surjective f.kerLift_injective⟩
+  refine ⟨(PrimeSpectrum.comap f).continuous, ?_, h1, hbij.2⟩
+  rw [PrimeSpectrum.isTopologicalBasis_basic_opens.isOpenMap_iff]
+  rintro - ⟨s, rfl⟩
+  obtain ⟨n, hn, r, hr⟩ := H s
+  have : (PrimeSpectrum.comap f) '' (PrimeSpectrum.basicOpen s) = PrimeSpectrum.basicOpen r := by
+    refine Set.preimage_injective.mpr hbij.2 ?_
+    rw [Set.preimage_image_eq _ hbij.1, ← PrimeSpectrum.basicOpen_pow _ n hn, ← hr]
+    rfl
+  rw [this]
+  exact PrimeSpectrum.isOpen_basicOpen
+
+variable {R S} in
+lemma PrimeSpectrum.isHomeomorph_comap_of_bijective {f : R →+* S} (hf : Function.Bijective f) :
+    IsHomeomorph (PrimeSpectrum.comap f) := by
```

Comment:
This is true for CommSemirings as well, see [here](https://github.com/leanprover-community/mathlib4/pull/20680).